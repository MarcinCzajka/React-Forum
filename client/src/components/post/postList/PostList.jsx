import React from 'react';
import { Grid } from "semantic-ui-react";
import ForumPostContainer from '../forumPost/ForumPostContainer';
import PostListPagination from './pagination/PostListPagination';
import { getForumPosts } from '../forumPost/forumPostLogic/forumPostApi';
import './PostList.scss';

class PostList extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            forumPosts: [],
            activePage: 1,
            totalPages: 1,
            postsLimit: 5,
            removedPostsCount: 0,

            windowWidth: window.innerWidth
        }
    }

    componentDidMount() {
        this.fetchForumPosts();
        window.addEventListener("resize", this.handleResize);
    }

    handleResize = () => {
        this.setState({windowWidth: window.innerWidth});
    }

    removeForumPostFromState = (key) => {
		//Function to be used in case Image is no longer available
        const forumPosts = this.state.forumPosts.filter(post => {
            return post.key !== key
        })

        const {postsLimit, removedPostsCount} = this.state;
        const totalPages =  Math.ceil((forumPosts.length - (removedPostsCount + 1)) / postsLimit);

        this.setState({forumPosts: forumPosts, removedPostsCount: removedPostsCount +1, totalPages: totalPages})
    }

    changePage = (e, { activePage }) => {
        this.setState({ activePage: activePage });
        this.fetchForumPosts(activePage);
    }

    fetchForumPosts = (pageNr = 1) => {
        const { postsLimit } = this.state;
        
        getForumPosts(pageNr, postsLimit)
            .then(res => {
                const forumPosts = res.rooms.map(item => {
                    return {...item, ...{key: item._id}}
                })

                const totalPages =  Math.ceil(res.totalCount / postsLimit);

                this.setState({forumPosts: forumPosts, totalPages: totalPages});
            })
    }

    render() {
        const forumPosts = this.state.forumPosts.map(room => {
            return (
                <Grid.Row key={room.key} centered>
                    <ForumPostContainer {...room} 
                        removeForumPost={this.removeForumPostFromState}
                    />
                </Grid.Row>
            );
        })

        const paginationProps = {
            activePage: this.state.activePage,
            totalPages: this.state.totalPages,
            windowWidth: this.state.windowWidth,
            changePage: this.changePage
        }
        
        return (
            <Grid id='postList' className='noMargin'>
                <PostListPagination 
                    {...paginationProps}
                />

                {forumPosts}

                <PostListPagination 
                    {...paginationProps}
                />
            </Grid>
        );
    }

}

export default PostList;