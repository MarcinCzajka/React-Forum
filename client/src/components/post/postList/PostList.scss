#postList {
    padding-top: 4em;
    padding-bottom:70px;
}