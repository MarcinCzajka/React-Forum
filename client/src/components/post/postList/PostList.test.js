import React from 'react';
import { shallow } from 'enzyme';
import { create } from "react-test-renderer";
import PostList from './PostList';

describe('PostListPagination component', () => {
    test('renders without crashing', () => {
        shallow(<PostList />);
    });

    test('match snapshot', () => {
        const component = create(<PostList />);
        expect(component).toMatchSnapshot();
    });

})