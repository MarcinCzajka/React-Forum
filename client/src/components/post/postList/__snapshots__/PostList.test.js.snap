// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListPagination component match snapshot 1`] = `
<div
  className="ui grid noMargin"
>
  
  
</div>
`;
